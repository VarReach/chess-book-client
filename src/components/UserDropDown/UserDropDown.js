import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import TokenService from '../../services/token-service';
import UserContext from '../../contexts/UserContext';

export default class DropDown extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showDropDown: false,
    };
    this.dropDown = React.createRef();
  }

  static contextType = UserContext;

  componentWillUnmount() {
    if (this.state.showDropDown) {
      document.removeEventListener('click', this.closeDropDown);
    }
  }

  toggleDropDown = (e) => {
    if (e && this.dropDown.current.contains(e.target)) {
      return;
    }
    this.state.showDropDown ? this.hideDropDown() : this.showDropDown();
  }

  showDropDown = () => {
    this.setState({ showDropDown: true }, () => {document.addEventListener('click', this.hideDropDown)});
  }

  hideDropDown = () => {
    this.setState({ showDropDown: false }, () => {document.removeEventListener('click', this.hideDropDown)});
  }

  render() {
    return (
      <div ref={this.dropDown} id="app__header-user-dropdown" className={"dropdown" + (!this.state.showDropDown ? ' hidden' : '')}>
        <ul>
        {TokenService.hasAuthToken()
            ? <li><Link to ="/" onClick={this.context.logout}><i className="fas fa-door-open"/>Logout</Link></li>
            : (
                <>
                  <li><Link to="/login"><i className="fas fa-sign-in-alt"/>Log in</Link></li>
                  <li><Link to="/register"><i className="fas fa-user-plus"/>Sign up</Link></li>
                </>
            )}
        </ul>
      </div>
    )
  }
}
