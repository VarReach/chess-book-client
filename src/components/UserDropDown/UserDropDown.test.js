import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import UserDropDown from './UserDropDown';

describe('UserDropDown component', () => {
  it('renders the dropdown when state.showDropDown is true', () => {
    const wrapper = mount(
      <MemoryRouter>
        <UserDropDown /> 
      </MemoryRouter>
    );
    wrapper.find(UserDropDown).instance().showDropDown();
    wrapper.update();
    expect(toJson(wrapper.find('#app__header-user-dropdown'))).toMatchSnapshot();
    wrapper.unmount();
  });
});