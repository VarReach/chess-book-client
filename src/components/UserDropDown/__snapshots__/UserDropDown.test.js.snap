// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDropDown component renders the dropdown when state.showDropDown is true 1`] = `
<div
  className="dropdown"
  id="app__header-user-dropdown"
>
  <ul>
    <li>
      <Link
        replace={false}
        to="/login"
      >
        <a
          href="/login"
          onClick={[Function]}
        >
          <i
            className="fas fa-sign-in-alt"
          />
          Log in
        </a>
      </Link>
    </li>
    <li>
      <Link
        replace={false}
        to="/register"
      >
        <a
          href="/register"
          onClick={[Function]}
        >
          <i
            className="fas fa-user-plus"
          />
          Sign up
        </a>
      </Link>
    </li>
  </ul>
</div>
`;
